feathr_project/feathr/_preprocessing_pyudf_manager.py

```diff
@@ -0,0 +1,151 @@
+import inspect
+import os
+from pathlib import Path
+from typing import List, Optional, Union
+import pickle
+from jinja2 import Template
+from feathr.source import HdfsSource
+
+# Some metadata that are only needed by Feathr
+FEATHR_PYSPARK_METADATA = 'generated_feathr_pyspark_metadata'
+# UDFs that are provided by users and persisted by Feathr into this file.
+# It will be uploaded into the Pyspark cluster as a dependency for execution
+FEATHR_CLIENT_UDF_FILE_NAME = 'client_udf_repo.py'
+# Pyspark driver code that is executed by the Pyspark driver
+FEATHR_PYSPARK_DRIVER_FILE_NAME = 'feathr_pyspark_driver.py'
+FEATHR_PYSPARK_DRIVER_TEMPLATE_FILE_NAME = 'feathr_pyspark_driver_template.py'
+# Feathr provided imports for pyspark UDFs all go here
+PROVIDED_IMPORTS = ['\nfrom pyspark.sql import SparkSession, DataFrame\n'] + \
+                   ['from pyspark.sql.functions import *\n']
+
+
+class _PreprocessingPyudfManager(object):
+    """This class manages Pyspark UDF preprocessing related artifacts, like UDFs from users, the pyspark_client etc.
+    """
+    def build_anchor_preprocessing_metadata(self, anchor_list, local_workspace_dir):
+        # feature names concatenated to UDF map
+        # for example, {'f1,f2,f3': my_udf1, 'f4,f5':my_udf2}
+        feature_names_to_func_mapping = {}
+        # features that have preprocessing defined. This is used to figure out if we need to kick off Pyspark
+        # preprocessing for requested features.
+        features_with_preprocessing = []
+        need_preprocessing = False
+        for anchor in anchor_list:
+            # only support batch source preprocessing for now.
```

Comment:
Why not other data sources? e.g. JDBC, as we just load them as DFs.

Reply:
currently all the JDBCs are implemented as HDFS sources.
typically it supports JDBCs and all HDFS sources.